stdlib/universe/double_exponential_moving_average.go

```diff
@@ -0,0 +1,406 @@
+
+	d := execute.NewDataset(id, mode, cache)
+	t := NewDEMATransformation(d, cache, alloc, s)
+	return t, d, nil
+}
+
+type DEMATransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+	alloc *memory.Allocator
+
+	n       int64
+	columns []string
+
+	ema1    *moving_average.ExponentialMovingAverage
+	ema2    *moving_average.ExponentialMovingAverage
+	i       []int
+	lastVal []interface{}
+}
+
+func NewDEMATransformation(d execute.Dataset, cache execute.TableBuilderCache, alloc *memory.Allocator, spec *DEMAProcedureSpec) *DEMATransformation {
+	return &DEMATransformation{
+		d:     d,
+		cache: cache,
+		alloc: alloc,
+
+		n:       spec.N,
+		columns: spec.Columns,
+	}
+}
+
+func (t *DEMATransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *DEMATransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("double exponential moving average found duplicate table with key: %v", tbl.Key())
+	}
+	cols := tbl.Cols()
+	doDEMA := make([]bool, len(cols))
+	for j, c := range cols {
+		found := false
+		for _, label := range t.columns {
+			if c.Label == label {
+				if c.Type != flux.TInt && c.Type != flux.TUInt && c.Type != flux.TFloat {
+					return fmt.Errorf("cannot take moving average of column %s (type %s)", c.Label, c.Type.String())
+				}
+				found = true
+				break
+			}
+		}
+
+		if found {
+			mac := c
+			mac.Type = flux.TFloat
+			_, err := builder.AddCol(mac)
+			if err != nil {
+				return err
+			}
+			doDEMA[j] = true
+		} else {
+			_, err := builder.AddCol(c)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	t.ema1 = moving_average.New(int(t.n), len(cols))
+	t.ema2 = moving_average.New(int(t.n), len(cols))
+
+	// Keeps track of current position and last value looked at for columns that are passed through
+	// Faster than calling ema.PassThrough twice
+	t.i = make([]int, len(cols))
+	t.lastVal = make([]interface{}, len(cols))
+
+	err := tbl.Do(func(cr flux.ColReader) error {
+		if cr.Len() == 0 {
+			return nil
+		}
+
+		// Start by doing the first EMA - we can't do both at once because of chunking
+		for j, c := range cols {
+
+			var err error
+			switch c.Type {
+			case flux.TBool:
+				err = t.passThrough(&moving_average.ArrayContainer{Array: cr.Bools(j)}, builder, j)
+			case flux.TInt:
+				err = t.doFirstEMA(&moving_average.ArrayContainer{Array: cr.Ints(j)}, builder, j, doDEMA[j])
+			case flux.TUInt:
+				err = t.doFirstEMA(&moving_average.ArrayContainer{Array: cr.UInts(j)}, builder, j, doDEMA[j])
+			case flux.TFloat:
+				err = t.doFirstEMA(&moving_average.ArrayContainer{Array: cr.Floats(j)}, builder, j, doDEMA[j])
+			case flux.TString:
+				err = t.passThrough(&moving_average.ArrayContainer{Array: cr.Strings(j)}, builder, j)
+			case flux.TTime:
+				err = t.passThroughTime(cr.Times(j), builder, j)
+			}
```

Comment:
Once we calculate the first EMA, we don't deal with chunks anymore. Is this a problem? Should data be kept in chunks?

Reply:
I think it's OK to not do the chunking.  it's not trivial to do that in this context.